cip25/rust/src/utils.rs

```diff
@@ -187,11 +192,401 @@ impl MiniMetadataDetails {
     }
 }
 
+#[derive(Copy, Clone, Debug, serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+pub enum CIP25Version {
+    V1,
+    V2,
+}
+
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+pub struct LabelMetadata {
+    pub nfts: BTreeMap<PolicyId, BTreeMap<AssetName, MetadataDetails>>,
+    pub version: CIP25Version,
+}
+
+impl LabelMetadata {
+    pub fn new(version: CIP25Version) -> Self {
+        Self {
+            nfts: BTreeMap::new(),
+            version,
+        }
+    }
+
+    pub fn set(&mut self, policy_id: PolicyId, asset_name: AssetName, details: MetadataDetails) -> Option<MetadataDetails> {
+        self.nfts.entry(policy_id).or_default().insert(asset_name, details)
+    }
+
+    pub fn get(&self, policy_id: &PolicyId, asset_name: &AssetName) -> Option<&MetadataDetails> {
+        self.nfts.get(policy_id)?.get(asset_name)
+    }
+}
+
+
+
+// serialization:
+
+impl cbor_event::se::Serialize for LabelMetadata {
+    fn serialize<'se, W: Write>(
+        &self,
+        serializer: &'se mut Serializer<W>,
+    ) -> cbor_event::Result<&'se mut Serializer<W>> {
+        match self.version {
+            CIP25Version::V1 => {
+                serializer.write_map(cbor_event::Len::Len(self.nfts.len() as u64))?;
+                for (policy_id, assets) in self.nfts.iter() {
+                    // hand-edit: write hex string
+                    serializer.write_text(policy_id.to_hex())?;
+                    serializer.write_map(cbor_event::Len::Len(assets.len() as u64))?;
+                    for (asset_name, details) in assets.iter() {
+                        // hand-edit: write hex string
```

Comment:
Oops. It's not hex actually. You can only encode string asset names in V1 AFAIK. What should we do here? Here are the first things that come to mind:
1) Throw an error. I am against this since, while cbor_event::Serialize::serialize() works with `Result`, in practice for us we should never throw an error as we use a vector serializer in most cases for our to/from bytes methods which assume that doing such will cause serialize() to never fail.
2) Encode as V2 as a fall-back if some of the asset names don't work with V1.
3) Get rid of pub access to `LabelMetadata`'s fields and have all setters throw an error if you either try to set the version to v1 and you have non-text asset names, or add an NFT with a non-text asset name while you're on v1.

Thoughts? I am leaning slightly towards 2) out of those options. Maybe there's another way too.

Reply:
Personally I feel like (3) seems the most intuitive behavior, but I agree it's kind of weird that the struct would contain this internal implicit constraint on how you can use it, which is why I have a slight preference to just wrap things similar to pallas-traverse instead of merging the structs like this

Reply:
We have #193 that tried wrapping instead of merging. The issue exists there too, you just shift where this restriction on v1 pops up in the API by doing that. In the wrapper there you'd instead have the setter have to return an error if you're trying to set a non-displayable asset name in V1, so any setting/creation using `AssetName` would have to return a `Result` to cover that possibility. I think we're stuck with either implicit v2 fallback or having an error _**somewhere**_ if we want to abstract the version diffs away (either with merging or wrappers).

And I don't think leaving things how they are without any version-independent API is a good idea because the part of the API that should matter least (just using policy ids/asset names which are concepts/types that already exist) takes up the majority of the API. What is `PolicyId` vs `PolicyIdV1` vs `PolicyIdV2`? (compared to reducing the burden on the user who would already know the `PolicyId` type from the rest of the library). The first version of the docs David made for this lib before I think give a good idea of what someone might think if they don't already know the version differences and try to make sense of the API. i.e. that it's confusing instead of having users just focus on the actual MetadataDetails that impact the NFT info and how to attach this to/parse from metadata. You could partially work around it with documentation in the first place but it's still nowhere near as clean.

Shall I try out 3) when I have time?

Reply:
> Shall I try out 3) when I have time?

Yeah I think fail-fast is better than the other options which delay the error occurring until later

Reply:
Oh yeah I'm definitely against 1) due to that (+ breaking our assumption we use in `to_bytes()` that serialization won't fail).

2) wouldn't have any error though, it would just switch to V2 format if you tried creating a V1 one with non-printable asset names. I guess it depends what you mean by error.

Reply:
The error is not when the silent switch happens - the error would be when you try and use it somewhere where you expected a V1 format and didn't realize it was silently converted to a V2 format and crashes your pipeline

Reply:
imo option 3 is the most intuitive for end-users. I think we should also add a comment about the differences between v1 and v2 near `LabelMetadata::new`